plugins/pulp_rpm/plugins/distributors/yum/publish.py

```diff
@@ -646,38 +658,36 @@ def _get_pkglists_to_publish(self, erratum_unit):
         :rtype: list of dicts
         """
         pkglists_to_publish = []
-
-        # If we can pull a repo_id off the conduit, use that to generate repo-specific nevra
-        if self.context.conduit and hasattr(self.context.conduit, 'repo_id'):
-            repo_unit_nevra = self.context.get_repo_unit_nevra(erratum_unit)
-        else:
-            repo_unit_nevra = None
+        repo_id = self.get_repo().id
+        repo_unit_nevra = self._get_repo_unit_nevra(repo_id)
 
         seen_pkglists = set()
         for pkglist in erratum_unit.pkglist:
-            packages = tuple(sorted(p['filename'] for p in pkglist['packages']))
-            if packages in seen_pkglists:
-                continue
-            seen_pkglists.add(packages)
-
-            if repo_unit_nevra is None:
-                pkglists_to_publish.append(pkglist)
-                continue
+            # Due to a quirk in how yum merges pkglists in errata that appear in multiple
+            # repos, the pkglist collection name needs to be unique per-repo when published.
+            # repo_id is already unique and presumably validated, so use that.
+            pkglist['name'] = repo_id
```

Comment:
As far as I can tell, the collection name is only used for the purposes of merging errata.

Looking at the two chunks of dtd in the yum code reveals these two things:
- There should be one and only one (so not zero, and not more than one) pkglist subelement attached to the "update" tag.[0]
- There can be multiple collections under a single pkglist.[1] yum does iterate over all collections within a pkglist, merging the packages within them, but only if the collection name is different, signifying that the collection is provided by a different repository.[2]

Despite the dtd specifying that there should be only one pkglist, yum does still loop over all available pkglist elements in a notice[3], storing the data for each in a list of parsed pkglist elements. This means that the collection name does indeed need to be unique for each collection in each errata in each published repository. So pulp either needs to merge all pkglists into a single collection per-repo, or we need to do a little more than just use 'repo_id' as the collection name to ensure the required uniqueness.

[0]: https://github.com/rpm-software-management/yum/blob/master/yum/update_md.py#L247-L249
[1]: https://github.com/rpm-software-management/yum/blob/master/yum/update_md.py#L314-L318
[2]: https://github.com/rpm-software-management/yum/blob/master/yum/update_md.py#L524-L529
[3]: https://github.com/rpm-software-management/yum/blob/master/yum/update_md.py#L255-L282